internal/mode/static/state/graph/route_common.go

```diff
@@ -0,0 +1,546 @@
+	validator validation.HTTPFieldsValidator,
+	httpRoutes map[types.NamespacedName]*v1.HTTPRoute,
+	grpcRoutes map[types.NamespacedName]*v1alpha2.GRPCRoute,
+	gatewayNsNames []types.NamespacedName,
+) map[RouteKey]*L7Route {
+	if len(gatewayNsNames) == 0 {
+		return nil
+	}
+
+	routes := make(map[RouteKey]*L7Route)
+
+	for _, ghr := range httpRoutes {
+		r := buildHTTPRoute(validator, ghr, gatewayNsNames)
+		if r != nil {
+			rk := RouteKey{
+				NamespacedName: client.ObjectKeyFromObject(ghr),
+				RouteType:      RouteTypeHTTP,
+			}
+			routes[rk] = r
+		}
+	}
+
+	for _, ghr := range grpcRoutes {
+		r := buildGRPCRoute(validator, ghr, gatewayNsNames)
+		if r != nil {
+			rk := RouteKey{
+				NamespacedName: client.ObjectKeyFromObject(ghr),
+				RouteType:      RouteTypeGRPC,
+			}
+			routes[rk] = r
+		}
+	}
+
+	return routes
+}
+
+func buildSectionNameRefs(
+	parentRefs []v1.ParentReference,
+	routeNamespace string,
+	gatewayNsNames []types.NamespacedName,
+) ([]ParentRef, error) {
+	sectionNameRefs := make([]ParentRef, 0, len(parentRefs))
+
+	type key struct {
+		gwNsName    types.NamespacedName
+		sectionName string
+	}
+	uniqueSectionsPerGateway := make(map[key]struct{})
+
+	for i, p := range parentRefs {
+		gw, found := findGatewayForParentRef(p, routeNamespace, gatewayNsNames)
+		if !found {
+			continue
+		}
+
+		var sectionName string
+		if p.SectionName != nil {
+			sectionName = string(*p.SectionName)
+		}
+
+		k := key{
+			gwNsName:    gw,
+			sectionName: sectionName,
+		}
+
+		if _, exist := uniqueSectionsPerGateway[k]; exist {
+			return nil, fmt.Errorf("duplicate section name %q for Gateway %s", sectionName, gw.String())
+		}
+		uniqueSectionsPerGateway[k] = struct{}{}
+
+		sectionNameRefs = append(sectionNameRefs, ParentRef{
+			Idx:     i,
+			Gateway: gw,
+		})
+	}
+
+	return sectionNameRefs, nil
+}
+
+func findGatewayForParentRef(
+	ref v1.ParentReference,
+	routeNamespace string,
+	gatewayNsNames []types.NamespacedName,
+) (gwNsName types.NamespacedName, found bool) {
+	if ref.Kind != nil && *ref.Kind != "Gateway" {
+		return types.NamespacedName{}, false
+	}
+	if ref.Group != nil && *ref.Group != v1.GroupName {
+		return types.NamespacedName{}, false
+	}
+
+	// if the namespace is missing, assume the namespace of the HTTPRoute
+	ns := routeNamespace
+	if ref.Namespace != nil {
+		ns = string(*ref.Namespace)
+	}
+
+	for _, gw := range gatewayNsNames {
+		if gw.Namespace == ns && gw.Name == string(ref.Name) {
+			return gw, true
+		}
+	}
+
+	return types.NamespacedName{}, false
+}
+
+func bindRoutesToListeners(
+	routes map[RouteKey]*L7Route,
+	gw *Gateway,
+	namespaces map[types.NamespacedName]*apiv1.Namespace,
+) {
+	if gw == nil {
+		return
+	}
+
+	for _, r := range routes {
+		bindRouteToListeners(r, gw, namespaces)
+	}
+}
+
+func bindRouteToListeners(
+	route *L7Route,
+	gw *Gateway,
+	namespaces map[types.NamespacedName]*apiv1.Namespace,
+) {
+	if !route.Attachable {
+		return
+	}
+
+	bindRoutes := func(r *L7Route) {
```

Comment:
Why this nested function?

Reply:
Ah, I think for cleanliness it was written that way. In your refactor, I don't think there's much value in it.

Reply:
I'd still ditch this nested function since it doesn't offer much value anymore.

Reply:
Sorry @sjberman I totally missed this in the last round, and I was mixed up about which nested function you were referring to 😅